File: src/main/java/org/wysko/midis2jam2/instrument/percussion/Timbales.java
package org.wysko.midis2jam2.instrument.percussion;

import com.jme3.math.Quaternion;
import com.jme3.scene.Node;
import com.jme3.scene.Spatial;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.instrument.Stick;
import org.wysko.midis2jam2.instrument.percussion.drumset.NonDrumSetPercussion;
import org.wysko.midis2jam2.instrument.percussion.drumset.StickDrum;
import org.wysko.midis2jam2.midi.MidiNoteOnEvent;

import java.util.List;
import java.util.stream.Collectors;

import static org.wysko.midis2jam2.Midis2jam2.rad;

/**
 * The timbales.
 */
public class Timbales extends NonDrumSetPercussion {
	
	private final List<MidiNoteOnEvent> lowTimbaleHits;
	
	private final List<MidiNoteOnEvent> highTimbaleHits;
	
	/**
	 * The Right hand node.
	 */
	private final Node highStickNode = new Node();
	
	/**
	 * The Left hand node.
	 */
	private final Node lowStickNode = new Node();
	
	/**
	 * The Left conga anim node.
	 */
	private final Node lowTimbaleAnimNode = new Node();
	
	/**
	 * The Right conga anim node.
	 */
	private final Node highTimbaleAnimNode = new Node();
	
	/**
	 * The low timbale.
	 */
	Spatial lowTimbale;
	
	/**
	 * The high timbale.
	 */
	Spatial highTimbale;
	
	/**
	 * Instantiates timbales.
	 *
	 * @param context the context
	 * @param hits    the hits
	 */
	public Timbales(Midis2jam2 context,
	                List<MidiNoteOnEvent> hits) {
		super(context, hits);
		
		lowTimbaleHits = hits.stream().filter(h -> h.note == 66).collect(Collectors.toList());
		highTimbaleHits = hits.stream().filter(h -> h.note == 65).collect(Collectors.toList());
		
		lowTimbale = context.loadModel("DrumSet_Timbale.obj", "DrumShell_Timbale.bmp");
		highTimbale = context.loadModel("DrumSet_Timbale.obj", "DrumShell_Timbale.bmp");
		highTimbale.setLocalScale(0.75f);
		
		lowTimbaleAnimNode.attachChild(lowTimbale);
		highTimbaleAnimNode.attachChild(highTimbale);
		
		Node lowTimbaleNode = new Node();
		lowTimbaleNode.attachChild(lowTimbaleAnimNode);
		Node highTimbaleNode = new Node();
		highTimbaleNode.attachChild(highTimbaleAnimNode);
		
		instrumentNode.attachChild(lowTimbaleNode);
		instrumentNode.attachChild(highTimbaleNode);
		
		lowStickNode.attachChild(context.loadModel("DrumSet_Stick.obj", "StickSkin.bmp"));
		highStickNode.attachChild(context.loadModel("DrumSet_Stick.obj", "StickSkin.bmp"));
		
		lowTimbaleAnimNode.attachChild(lowStickNode);
		highTimbaleAnimNode.attachChild(highStickNode);
		
		lowTimbaleNode.setLocalTranslation(-45.9f, 50.2f, -59.1f);
		lowTimbaleNode.setLocalRotation(new Quaternion().fromAngles(rad(32), rad(56.6), rad(-2.6)));
		
		highTimbaleNode.setLocalTranslation(-39, 50.1f, -69.7f);
		highTimbaleNode.setLocalRotation(new Quaternion().fromAngles(rad(33.8), rad(59.4), rad(-1.8)));
		
		lowStickNode.setLocalTranslation(0, 0, 10);
		highStickNode.setLocalTranslation(0, 0, 10);
	}
	
	@Override
	public void tick(double time, float delta) {
		super.tick(time, delta);
		
		Stick.StickStatus statusLow = Stick.handleStick(context, lowStickNode, time, delta, lowTimbaleHits, Stick.STRIKE_SPEED, Stick.MAX_ANGLE);
		Stick.StickStatus statusHigh = Stick.handleStick(context, highStickNode, time, delta, highTimbaleHits, Stick.STRIKE_SPEED, Stick.MAX_ANGLE);
		
		if (statusLow.justStruck()) {
			MidiNoteOnEvent strike = statusLow.getStrike();
			assert strike != null;
			StickDrum.recoilDrum(lowTimbaleAnimNode, true, strike.velocity, delta);
		} else {
			StickDrum.recoilDrum(lowTimbaleAnimNode, false, 0, delta);
		}
		
		if (statusHigh.justStruck()) {
			MidiNoteOnEvent strike = statusHigh.getStrike();
			assert strike != null;
			StickDrum.recoilDrum(highTimbaleAnimNode, true, strike.velocity, delta);
		} else {
			StickDrum.recoilDrum(highTimbaleAnimNode, false, 0, delta);
		}
	}
}

File: app/src/commonMain/kotlin/org/wysko/midis2jam2/instrument/family/percussion/Timbales.kt
/*
 * Copyright (C) 2025 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.percussion

import com.jme3.math.Quaternion
import com.jme3.scene.Node
import org.wysko.kmidi.midi.event.NoteEvent
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.instrument.algorithmic.StickType
import org.wysko.midis2jam2.instrument.algorithmic.Striker
import org.wysko.midis2jam2.util.Utils.rad
import org.wysko.midis2jam2.world.modelD
import kotlin.time.Duration

/** The Timbales. */
class Timbales(
    context: Midis2jam2,
    highHits: MutableList<NoteEvent.NoteOn>,
    lowHits: MutableList<NoteEvent.NoteOn>,
) : AuxiliaryPercussion(context, (highHits + lowHits).sortedBy { it.tick }.toMutableList()) {
    /** The left timbale anim node. */
    private val lowTimbaleAnimNode =
        Node().also {
            it.attachChild(context.modelD("DrumSet_Timbale.obj", "DrumShell_Timbale.bmp"))
        }

    /** The right timbale anim node. */
    private val highTimbaleAnimNode =
        Node().also {
            it.attachChild(
                context.modelD("DrumSet_Timbale.obj", "DrumShell_Timbale.bmp").apply {
                    scale(0.75f)
                },
            )
        }

    private val lowStick =
        Striker(
            context = context,
            lowHits,
            stickModel = StickType.DRUM_SET_STICK,
        ).apply {
            setParent(lowTimbaleAnimNode)
            node.move(0f, 0f, 10f)
        }

    private val highStick =
        Striker(
            context = context,
            highHits,
            stickModel = StickType.DRUM_SET_STICK,
        ).apply {
            setParent(highTimbaleAnimNode)
            node.move(0f, 0f, 10f)
        }

    init {
        // Low timbale node
        Node().apply {
            attachChild(lowTimbaleAnimNode)
            setLocalTranslation(-45.9f, 50.2f, -59.1f)
            localRotation = Quaternion().fromAngles(rad(32.0), rad(56.6), rad(-2.6))

            geometry.attachChild(this)
        }

        // High timbale node
        Node().apply {
            attachChild(highTimbaleAnimNode)
            setLocalTranslation(-39f, 50.1f, -69.7f)
            localRotation = Quaternion().fromAngles(rad(33.8), rad(59.4), rad(-1.8))

            geometry.attachChild(this)
        }
    }

    override fun tick(
        time: Duration,
        delta: Duration,
    ) {
        super.tick(time, delta)

        recoilDrum(lowTimbaleAnimNode, lowStick.tick(time, delta).velocity, delta)
        recoilDrum(highTimbaleAnimNode, highStick.tick(time, delta).velocity, delta)
    }
}
